src/quota_aggregator_impl_test.cc

```diff
@@ -0,0 +1,428 @@
+  EXPECT_TRUE(MessageDifferencer::Equals(response2, pass_response2_));
+
+  // check flushed out list is empty
+  EXPECT_EQ(flushed_.size(), 0);
+
+  std::this_thread::sleep_for(std::chrono::milliseconds(60));
+
+  // expire request1
+  EXPECT_OK(aggregator_->Flush());
+  EXPECT_EQ(flushed_.size(), 1);
+
+  EXPECT_TRUE(flushed_[0].has_allocate_operation());
+  quota_metrics = covertMetricSets(flushed_[0].allocate_operation());
+  expected_costs = {{"metric_first", 1}, {"metric_second", 1}};
+  ASSERT_EQ(expected_costs, quota_metrics);
+
+  std::this_thread::sleep_for(std::chrono::milliseconds(60));
+
+  // expire request2
+  EXPECT_OK(aggregator_->Flush());
+  EXPECT_EQ(flushed_.size(), 2);
+
+  EXPECT_TRUE(flushed_[1].has_allocate_operation());
+  expected_costs = {{"metric_first", 2}, {"metric_second", 3}};
+  quota_metrics = covertMetricSets(flushed_[1].allocate_operation());
+  ASSERT_EQ(quota_metrics, expected_costs);
+
+  // expire temporary elements for refreshment from the cache
+  std::this_thread::sleep_for(std::chrono::milliseconds(110));
+  EXPECT_OK(aggregator_->Flush());
+  EXPECT_EQ(flushed_.size(), 2);
+
+  // lookup request 1 again
+  AllocateQuotaResponse response3;
+  EXPECT_ERROR_CODE(Code::NOT_FOUND, aggregator_->Quota(request1_, &response3));
+  EXPECT_TRUE(MessageDifferencer::Equals(response3, empty_response_));
+}
+
+TEST_F(QuotaAggregatorImplTest, TestCacheRefreshOneAggregated) {
+  std::set<std::pair<std::string, int>> quota_metrics;
+  std::set<std::pair<std::string, int>> expected_costs;
+
+  AllocateQuotaResponse response1;
+  AllocateQuotaResponse response2;
+
+  // insert request 1
+  EXPECT_ERROR_CODE(Code::NOT_FOUND, aggregator_->Quota(request1_, &response1));
+  EXPECT_OK(aggregator_->CacheResponse(request1_, pass_response1_));
+
+  std::this_thread::sleep_for(std::chrono::milliseconds(50));
+
+  // insert request 2
+  EXPECT_ERROR_CODE(Code::NOT_FOUND, aggregator_->Quota(request2_, &response2));
+  EXPECT_OK(aggregator_->CacheResponse(request2_, pass_response2_));
+
+  // aggregate request 1
+  EXPECT_OK(aggregator_->Quota(request1_, &response1));
+
+  // check responses from cache
+  EXPECT_TRUE(MessageDifferencer::Equals(response1, pass_response1_));
+  EXPECT_TRUE(MessageDifferencer::Equals(response2, empty_response_));
+
+  // nothing flushed out yet
+  EXPECT_EQ(flushed_.size(), 0);
+
+  std::this_thread::sleep_for(std::chrono::milliseconds(60));
+
+  // expire request1
+  EXPECT_OK(aggregator_->Flush());
+
+  // check request1 has flushed out
+  EXPECT_EQ(flushed_.size(), 1);
+
+  // verify flushed out 1
+  expected_costs = {{"metric_first", 1}, {"metric_second", 1}};
+  quota_metrics = covertMetricSets(flushed_[0].allocate_operation());
+
+  std::this_thread::sleep_for(std::chrono::milliseconds(60));
+
+  // expire request2
+  EXPECT_OK(aggregator_->Flush());
+
+  // request2 hasn't been aggregated. No need to flushed out
+  EXPECT_EQ(flushed_.size(), 1);
```

Comment:
please add comment
// the second cached item did not have aggregated quota, when it is flushed, it was dropped.